### PR TITLE
[Network Modification] Move method returning feeder by connectable in TopologyModificationUtils from pypowsybl 

### PR DESCRIPTION
**Pull request by colinepiloquet**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Moves code from pypowsybl to core and refactor, it makes more sense that it's there with all the other methods linked to topology modification. The rest of the code is also refactored. 
This creates a method returning the feeders by connectable in TopologyModificationUtil.

### CODE REVIEW

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 1]
```diff
@@ -439,12 +439,14 @@ public static Set<Integer> getFeederPositions(VoltageLevel voltageLevel) {
      */
     public static Map<String, List<Integer>> getFeederPositionsByConnectable(VoltageLevel voltageLevel) {
         Map<String, List<Integer>> feederPositionsOrders = new HashMap<>();
-        voltageLevel.getConnectables().forEach(connectable -> {
-            ConnectablePosition<?> position = (ConnectablePosition<?>) connectable.getExtension(ConnectablePosition.class);
-            if (position != null) {
-                List<Integer> orders = getOrderPositions(position, voltageLevel, connectable, false, Reporter.NO_OP);
-                feederPositionsOrders.put(connectable.getId(), orders);
+        Map<String, List<ConnectablePosition.Feeder>> feedersByConnectable = getFeedersByConnectable(voltageLevel);
+        feedersByConnectable.forEach((k, v) -> {
```
Comment (miovd) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:443]: You can directly do `getFeedersByConnectable(voltageLevel).forEach(...` without using an intermediary variable.

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 2]
```diff
@@ -464,13 +466,52 @@ private static void addOrderPositions(Connectable<?> connectable, VoltageLevel v
         }
     }
 
+    /**
+     * Utility method to get all the feeder on a voltage level by connectable.
```
Comment (miovd) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:470]: Typo `feeders`

**Review by miovd — COMMENTED** [review]
Small comments but mostly okay for me!

**Review by miovd — APPROVED** [review]

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 3]
```diff
@@ -464,13 +465,52 @@ private static void addOrderPositions(Connectable<?> connectable, VoltageLevel v
...
+        } else {
+            LOGGER.error("Given connectable not supported: {}", connectable.getClass().getName());
+            connectableNotSupported(reporter, connectable);
+            if (throwException) {
+                throw new AssertionError("Given connectable not supported: " + connectable.getClass().getName());
+            }
+        }
+        List<Integer> orders = new ArrayList<>();
+        feeders.forEach(feeder -> feeder.getOrder().ifPresent(orders::add));
+        checkConnectableInVoltageLevel(orders, voltageLevel, connectable, throwException, reporter);
```
Comment (flo-dup) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:500]: In this method we don't really check that the connectable is in voltage level, in fact now that we moved it we also get this error if the connectable is not supported (well so far that's not possible but still). I'd rather check earlier if connectable is in voltage level by just doing `connectable.getTerminals().stream().anyMatch(t -> t.getVoltageLevel() == voltageLevel)`, instead of supposing we can see this in orders.

**Review by flo-dup — CHANGES_REQUESTED** [review]

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java (hunk 3 above)
Comment (miovd) [reply iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:500]: I agree, if you want to check if orders are empty, rename this method `checkOrdersExist`.
Before that, you can check that the connectable is in the voltage level as florian proposed.

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 4]
```diff
@@ -464,13 +465,52 @@ private static void addOrderPositions(Connectable<?> connectable, VoltageLevel v
...
+        } else if (connectable instanceof Branch) {
+            feeders = getBranchFeeders(position, voltageLevel, (Branch<?>) connectable);
+        } else if (connectable instanceof ThreeWindingsTransformer) {
+            feeders = get3wtFeeders(position, voltageLevel, (ThreeWindingsTransformer) connectable);
+        } else {
+            LOGGER.error("Given connectable not supported: {}", connectable.getClass().getName());
+            connectableNotSupported(reporter, connectable);
+            if (throwException) {
+                throw new AssertionError("Given connectable not supported: " + connectable.getClass().getName());
+            }
```
Comment (miovd) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:496]: Return `Collections.emptyList()` here? We have to stop the method here if not supported.

**Review by miovd — APPROVED** [review]

**Comment by miovd** [comment]
More tests of situations returning an empty list should be added.

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 5]
```diff
@@ -477,13 +478,61 @@ private static void addOrderPositions(Connectable<?> connectable, VoltageLevel v
...
+            if (position != null) {
+                List<ConnectablePosition.Feeder> feeder = getFeeders(position, voltageLevel, connectable, false, Reporter.NO_OP);
+                feedersByConnectable.put(connectable.getId(), feeder);
+            }
+        });
+        return feedersByConnectable;
+    }
+
     private static List<Integer> getOrderPositions(ConnectablePosition<?> position, VoltageLevel voltageLevel, Connectable<?> connectable, boolean throwException, Reporter reporter) {
+        if (connectable.getTerminals().stream().noneMatch(t -> t.getVoltageLevel() == voltageLevel)) {
```
Comment (flo-dup) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:497]: as discussed, this check is after all not necessary as this private method is always called on the connectables of the given voltage level

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 6]
```diff
@@ -494,47 +543,41 @@ private static List<Integer> getOrderPositions(ConnectablePosition<?> position,
...
+            Optional.ofNullable(position.getFeeder3()).ifPresent(feeders::add);
         }
-        checkConnectableInVoltageLevel(orders, voltageLevel, twt, throwException, reporter);
-        Collections.sort(orders);
-        return orders;
+        return feeders;
     }
 
-    private static void checkConnectableInVoltageLevel(List<Integer> orders, VoltageLevel voltageLevel, Connectable<?> connectable, boolean throwException, Reporter reporter) {
+    private static void checkOrdersExist(List<Integer> orders, Connectable<?> connectable, boolean throwException, Reporter reporter) {
```
Comment (flo-dup) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:575]: This is called when collecting all the orders of a slice or of a voltage level, I don't think this check is needed. We could rather do a higher level check on whether there's no orders at all on the slice / voltage level

File: iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java [hunk 7]
```diff
@@ -494,47 +543,41 @@ private static List<Integer> getOrderPositions(ConnectablePosition<?> position,
...
         if (orders.isEmpty()) {
-            LOGGER.error("Given connectable {} not in voltageLevel {}", connectable.getId(), voltageLevel.getId());
-            connectableNotInVoltageLevel(reporter, connectable, voltageLevel);
+            LOGGER.error("Given connectable {} has no orders", connectable.getId());
```
Comment (flo-dup) [inline iidm/iidm-modification/src/main/java/com/powsybl/iidm/modification/topology/TopologyModificationUtils.java:577]: Just if you move this message at the voltage level / slice level, you should add the "position" word, as no orders alone sounds like no command!

**Review by flo-dup — APPROVED** [review]

**Comment by miovd** [comment]
@colinepiloquet There are some failure tests, can you check?

**Comment by colinepiloquet** [comment]
I fixed the test! It should be good now.

File: iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/TopologyModificationUtilsTest.java [hunk 8]
```diff
@@ -71,13 +71,13 @@ public void testGetFeedersByConnectable() {
...
         assertEquals("trf61", feeder.get(0).getName().orElse(null));
-        assertEquals(Optional.of(50), feeder.get(0).getOrder());
+        assertEquals(50, feeder.get(0).getOrder().orElse(0), 0);
         Map<String, List<ConnectablePosition.Feeder>> feedersVl3 = getFeedersByConnectable(network.getVoltageLevel("vl3"));
         assertFalse(feedersVl3.isEmpty());
         assertTrue(feedersVl3.containsKey("trf7"));
         List<ConnectablePosition.Feeder> trf7Vl2 = feedersVl3.get("trf7");
         assertEquals(1, trf7Vl2.size());
-        assertEquals(Optional.of(30), trf7Vl2.get(0).getOrder());
+        assertEquals(30, trf7Vl2.get(0).getOrder().orElse(0),0);
```
Comment (miovd) [inline iidm/iidm-modification/src/test/java/com/powsybl/iidm/modification/topology/TopologyModificationUtilsTest.java:80]: Missing whitespace :)
